pkg/controller/atlasproject/custom_roles.go

```diff
@@ -0,0 +1,201 @@
+package atlasproject
+
+import (
+	"context"
+
+	"github.com/mongodb/mongodb-atlas-kubernetes/pkg/api/v1/status"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-cmp/cmp/cmpopts"
+
+	v1 "github.com/mongodb/mongodb-atlas-kubernetes/pkg/api/v1"
+	"github.com/mongodb/mongodb-atlas-kubernetes/pkg/controller/workflow"
+)
+
+func ensureCustomRoles(ctx *workflow.Context, projectID string, project *v1.AtlasProject) (result workflow.Result) {
+	isFailure := func(err error) bool {
+		if err != nil {
+			result = workflow.Terminate(workflow.ProjectCustomRolesReady, err.Error())
+			ctx.SetConditionFromResult(status.ProjectCustomRolesReadyType, result)
+
+			return true
+		}
+
+		return false
+	}
+
+	currentCustomRoles, err := fetchCustomRoles(ctx, projectID)
+	if isFailure(err) {
+		return
+	}
+
+	err = deleteCustomRoles(ctx, projectID, currentCustomRoles, project.Spec.CustomRoles)
+	if isFailure(err) {
+		return
+	}
+
+	err = updateCustomRoles(ctx, projectID, currentCustomRoles, project.Spec.CustomRoles)
+	if isFailure(err) {
+		return
+	}
+
+	err = createCustomRoles(ctx, projectID, currentCustomRoles, project.Spec.CustomRoles)
+	if isFailure(err) {
+		return
+	}
+
+	ctx.SetConditionTrue(status.ProjectCustomRolesReadyType)
+
+	if len(project.Spec.CustomRoles) == 0 {
+		ctx.UnsetCondition(status.ProjectCustomRolesReadyType)
+	}
+
+	return workflow.OK()
+}
+
+func fetchCustomRoles(ctx *workflow.Context, projectID string) ([]v1.CustomRole, error) {
+	data, _, err := ctx.Client.CustomDBRoles.List(context.Background(), projectID, nil)
+	if err != nil {
+		return nil, err
```

Comment:
It might be worth wrapping up errors because they will be in status. And without context, it might be hard to understand what exactly happened.